README.md
# bee-auto-complete

[![npm version](https://img.shields.io/npm/v/bee-auto-complete.svg)](https://www.npmjs.com/package/bee-auto-complete)
[![Build Status](https://img.shields.io/travis/tinper-bee/bee-auto-complete/master.svg)](https://travis-ci.org/tinper-bee/bee-auto-complete)
[![Coverage Status](https://coveralls.io/repos/github/tinper-bee/bee-auto-complete/badge.svg?branch=master)](https://coveralls.io/github/tinper-bee/bee-auto-complete?branch=master)
[![devDependency Status](https://img.shields.io/david/dev/tinper-bee/bee-auto-complete.svg)](https://david-dm.org/tinper-bee/bee-auto-complete#info=devDependencies)
[![NPM downloads](http://img.shields.io/npm/dm/bee-auto-complete.svg?style=flat)](https://npmjs.org/package/bee-auto-complete)
[![Average time to resolve an issue](http://isitmaintained.com/badge/resolution/tinper-bee/bee-auto-complete.svg)](http://isitmaintained.com/project/tinper-bee/bee-auto-complete "Average time to resolve an issue")
[![Percentage of issues still open](http://isitmaintained.com/badge/open/tinper-bee/bee-auto-complete.svg)](http://isitmaintained.com/project/tinper-bee/bee-auto-complete "Percentage of issues still open")


react bee-auto-complete component for tinper-bee

根据用户的输入，进行自动匹配显示列表，用户可以进行按需选择自己需要的内容在输入框展示

## 使用方法

```
class Demo1 extends Component {
    constructor(props) {
        super(props);
        this.state = {
            value: "",
            options: ['10000(博宇)', '10001(波波)', '10002(李刚)'],
            placeholder: "查找关键字,请输入1",
            disabled: false
        }
        this.onFormChange = this.onFormChange.bind(this);
    }
    onFormChange(value) {
        this.setState({
            value: value
        })

    }
    render() {
        let { value, options, placeholder,disabled} = this.state;
        return (
            <div className="demo" style={{ "marginBottom": "90px" }}>
                <AutoComplete
                    value={value}
                    disabled={disabled}
                    options={options}
                    placeholder={placeholder}
                    onValueChange={value => this.onFormChange(value)}
                />

            </div>
        )
    }
}

```



## API
|参数|说明|类型|默认值|
|:---|:-----|:----|:------|
|value|传入的输入框的值|any|-|
|options|自动匹配列表|array|[]|
|placeholder|输入框的默认提示|string|-|
|show|控制自动匹配列表的展开和隐藏|boolean|false|
|onChange| input 的 value 变化时，调用此函数|handleChange(value)|-|
|onKeyDown| 监听input的键盘事件，调用此函数|handleKeyDown()|-|
|onBlur| input 的失去焦点时，调用此函数|handLeBlur()|-|
|onMouseEnter| 自动匹配列表随着光标选中变化时候，调用此函数|handleEnter(index)|-|
|onClick| 自动匹配列表某一项列表被点击的时候，调用此函数|handleChangeList(item)|-|

#### 开发调试

```sh
$ npm install -g bee-tools
$ git clone https://github.com/tinper-bee/bee-auto-complete
$ cd bee-auto-complete
$ npm install
$ npm run dev
```


CHANGELOG.md
<a name="1.1.1"></a>
## [1.1.1](https://github.com/tinper-bee/bee-auto-complete/compare/v1.1.0...v1.1.1) (2019-03-01)



<a name="1.1.0"></a>
# [1.1.0](https://github.com/tinper-bee/bee-auto-complete/compare/v1.0.4...v1.1.0) (2019-02-25)



<a name="1.0.4"></a>
## [1.0.4](https://github.com/tinper-bee/bee-auto-complete/compare/70333a7...v1.0.4) (2018-03-08)


### Bug Fixes

* 修改样式规范，修改依赖引入 ([70333a7](https://github.com/tinper-bee/bee-auto-complete/commit/70333a7))
* 修改示例 ([eb6ba30](https://github.com/tinper-bee/bee-auto-complete/commit/eb6ba30))


### Features

* **AutoComplete.js:** 支持options参数动态变更 ([4cec152](https://github.com/tinper-bee/bee-auto-complete/commit/4cec152))





src/AutoComplete.js
import React, { Component } from 'react';
import classnames from 'classnames';
import FormControl from 'bee-form-control';
import PropTypes from 'prop-types';

const propTypes = {
    value: PropTypes.any,
    options: PropTypes.array,
    onChange: PropTypes.func
};
const defaultProps = {
    value: "",
    options: [],
    clsPrefix: 'u-autocomplete'
};

class AutoComplete extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            show: false, //控制自动匹配列表的显示与隐藏
            displayValue: '',
            activeItemIndex: -1,
            options: props.options,
            value: props.value,
            placeholder: props.placeholder
        };
        this.handleKeyDown = this.handleKeyDown.bind(this);
        this.handleLeave = this.handleLeave.bind(this);
        this.handleEnter = this.handleEnter.bind(this);
        this.handleChangeList = this.handleChangeList.bind(this);
        this.moveItem = this.moveItem.bind(this);
        this.handLeBlur = this.handLeBlur.bind(this);
    }
    componentWillReceiveProps(props) {
        if ('value' in props) {
            let value = props.value;
            this.setState({
                value: value
            })
        }
        if ('options' in props) {
            let options = props.options;
            this.setState({
                options: options
            })
        }
    }

    handleChange(value) {
        let arr = [];
        let valueArr = this.props.options;
        valueArr.sort();
        if (value.replace(/(^\s*)|(\s*$)/g, '') == "") {
            this.setState({
                value: "",
                activeItemIndex: -1,
                show: false
            })
            this.props.onValueChange(value);
            return;
        }

        for (var i = 0; i < valueArr.length; i++) {
            if (valueArr[i].indexOf(value) != -1) {
                arr.push(valueArr[i]);
            }
        }
        this.setState({
            options: arr,
            show: true,
            activeItemIndex: -1,
            displayValue: '',
            value: value
        })
        this.props.onValueChange(value);
    }
    /**
     * 自动匹配的列表被选中其中某一个
     * @param {*} value 
     */
    handleChangeList(value) {
        this.setState({
            show: false,
            displayValue: ''
        })
        this.props.onValueChange(value);
    }

    handleKeyDown(e) {
        const { activeItemIndex } = this.state;
        const { options } = this.props;

        switch (e.keyCode) {
            // 13为回车键的键码（keyCode）
            case 13: {
                this.setState({
                    show: false
                })
                break;
            }
            // 38为上方向键，40为下方向键
            case 38:
            case 40: {
                e.preventDefault();
                // 使用moveItem方法对更新或取消选中项
                this.moveItem(e.keyCode === 38 ? 'up' : 'down');
                break;
            }
        }
    }
    
    moveItem(direction) {
        const { activeItemIndex, options } = this.state;
        const lastIndex = options.length - 1;
        let newIndex = -1;

        // 计算新的activeItemIndex
        if (direction === 'up') {
            if (activeItemIndex === -1) {
                // 如果没有选中项则选择最后一项
                newIndex = lastIndex;
            } else {
                newIndex = activeItemIndex - 1;
            }
        } else {
            if (activeItemIndex < lastIndex) {
                newIndex = activeItemIndex + 1;
            }
        }

        // 获取新的displayValue
        let newDisplayValue = '';
        if (newIndex >= 0) {
            newDisplayValue = options[newIndex];
        }

        // 更新状态
        this.setState({
            displayValue: newDisplayValue,
            activeItemIndex: newIndex
        });
    }

    handleEnter(index) {
        const currentItem = this.props.options[index];
        this.setState({ activeItemIndex: index, displayValue: currentItem });
    }

    handleLeave() {
        this.setState({ activeItemIndex: -1, displayValue: '' });
    }
    handLeBlur() {
        this.setState({
            show: false
        });
    }
    render() {
        const { show, displayValue, activeItemIndex, options, value, placeholder } = this.state;
        const { disabled, clsPrefix } = this.props;
        return (
            <div className={classnames(clsPrefix,this.props.className)}>
                <FormControl
                    value={displayValue || value}
                    disabled={disabled }
                    onChange={(value) => {this.handleChange(value)}}
                    onKeyDown={this.handleKeyDown}
                    placeholder={placeholder}
                    onBlur={this.handLeBlur}
                />
                {show && options.length > 0 && (
                    <ul className={`${clsPrefix}-options`} onMouseLeave={this.handleLeave}>
                        {
                            options.map((item, index) => {
                                return (
                                    <li
                                        key={index}
                                        className={index === activeItemIndex ? "active" : ''}
                                        onMouseEnter={() => this.handleEnter(index)}
                                        onClick={() => this.handleChangeList(item)}
                                    >
                                        {item.text || item}
                                    </li>
                                );
                            })
                        }
                    </ul>
                )}
            </div>
        );
    }
}


AutoComplete.propTypes = propTypes;
AutoComplete.defaultProps = defaultProps;
export default AutoComplete;

demo/demolist/Demo1.js
/**
 *
 * @title 根据输入框的内容，进行自动匹配列表显示
 * @description 用户可以按需进行选择，支持光标操作，回车事件
 *
 */
import React, { Component } from "react";
import AutoComplete from "../../src";

class Demo1 extends Component {
  constructor(props) {
    super(props);
    this.state = {
      value: "",
      options: ["10000(博宇)", "10001(波波)", "10002(李刚)"],
      placeholder: "查找关键字,请输入1",
      disabled: false
    };
  }
  onFormChange = value => {
    this.setState({
      value: value
    });
  };
  render() {
    let { value, options, placeholder, disabled } = this.state;
    return (
      <div className="demo" style={{ marginBottom: "90px" }}>
        <AutoComplete
          value={value}
          disabled={disabled}
          options={options}
          placeholder={placeholder}
          onValueChange={value => this.onFormChange(value)}
        />
      </div>
    );
  }
}

export default Demo1;


demo/demolist/Demo2.js

/**
*
* @title 根据输入框的内容，进行自动匹配列表显示
* @description 用户可以按需进行选择，支持光标操作，回车事件
*
*/
import React, { Component } from 'react';
import AutoComplete from '../../src'

class Demo2 extends Component {
    constructor(props) {
        super(props);
        this.state = {
            value: "",
            options: ['a', 'aa', 'aaa', 'aaaa'],
            placeholder: "查找关键字,请输入a",
            disabled:false
        }
        this.onFormChange = this.onFormChange.bind(this);
    }
    onFormChange(value) {
        this.setState({
            value: value
        })

    }
    render() {
        let { value, options, placeholder,disabled} = this.state;
        return (
            <div className="demo" style={{ "marginBottom": "110px" }}>
                <AutoComplete
                    value={value}
                    disabled={disabled}
                    options={options}
                    placeholder={placeholder}
                    onValueChange={value => this.onFormChange(value)}
                />

            </div>
        )
    }
}

export default Demo2;

demo/index-demo-base.js

import { Con, Row, Col } from 'bee-layout';
import { Panel } from 'bee-panel';
import Button from 'bee-button';
import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import AutoComplete from '../src';


const CARET = <i className="uf uf-arrow-down"></i>;

const CARETUP = <i className="uf uf-arrow-up"></i>;


{demolist}

class Demo extends Component {
    constructor(props){
        super(props);
        this.state = {
            open: false
        }
        this.handleClick = this.handleClick.bind(this);
    }
    handleClick() {
        this.setState({ open: !this.state.open })
    }

    render () {
        const { title, example, code, desc  } = this.props;
        let caret = this.state.open ? CARETUP : CARET;
        let text = this.state.open ? "隐藏代码" : "查看代码";

        const footer = (
            <Button shape="block" onClick={ this.handleClick }>
                { caret }
                { text }
            </Button>
        );
        return (
            <Col md={12} >
                <h3>{ title }</h3>
                <p>{ desc }</p>
                <Panel collapsible headerContent expanded={ this.state.open } colors='bordered' header={ example } footer={footer} footerStyle = {{padding: 0}}>
                    <pre><code className="hljs javascript">{ code }</code></pre>
                </Panel>
            </Col>
        )
    }
}

class DemoGroup extends Component {
    constructor(props){
        super(props)
    }
    render () {
        return (
                <Row>
                    {DemoArray.map((child,index) => {

                        return (
                            <Demo example= {child.example} title= {child.title} code= {child.code} desc= {child.desc} key= {index}/>
                        )

                    })}
                </Row>
        )
    }
}

ReactDOM.render(<DemoGroup/>, document.getElementById('tinperBeeDemo'));


demo/index.js
import {Col, Row} from 'bee-layout';
import {Panel} from 'bee-panel';
import Button from 'bee-button';
import React, {Component} from 'react';
import ReactDOM from 'react-dom';


const CARET = <i className="uf uf-arrow-down"></i>;

const CARETUP = <i className="uf uf-arrow-up"></i>;


var Demo1 = require("./demolist/Demo1");var Demo2 = require("./demolist/Demo2");var DemoArray = [{"example":<Demo1 />,"title":" 根据输入框的内容，进行自动匹配列表显示","code":"/**\n *\n * @title 根据输入框的内容，进行自动匹配列表显示\n * @description 用户可以按需进行选择，支持光标操作，回车事件\n *\n */\nimport React, { Component } from \"react\";\nimport { AutoComplete } from 'tinper-bee';\n\nclass Demo1 extends Component {\n  constructor(props) {\n    super(props);\n    this.state = {\n      value: \"\",\n      options: [\"10000(博宇)\", \"10001(波波)\", \"10002(李刚)\"],\n      placeholder: \"查找关键字,请输入1\",\n      disabled: false\n    };\n  }\n  onFormChange = value => {\n    this.setState({\n      value: value\n    });\n  };\n  render() {\n    let { value, options, placeholder, disabled } = this.state;\n    return (\n      <div className=\"demo\" style={{ marginBottom: \"90px\" }}>\n        <AutoComplete\n          value={value}\n          disabled={disabled}\n          options={options}\n          placeholder={placeholder}\n          onValueChange={value => this.onFormChange(value)}\n        />\n      </div>\n    );\n  }\n}\n\n\n","desc":" 用户可以按需进行选择，支持光标操作，回车事件"},{"example":<Demo2 />,"title":" 动态改变options参数","code":"/**\n *\n * @title 动态改变options参数\n * @description 用户可以按需进行选择，支持光标操作，回车事件\n *\n */\nimport React, { Component } from \"react\";\nimport { AutoComplete } from 'tinper-bee';\n\nclass Demo2 extends Component {\n  constructor(props) {\n    super(props);\n    this.state = {\n      value: \"\",\n      options: [],\n      placeholder: \"查找关键字,请输入a\",\n      disabled: false\n    };\n  }\n  onFormChange = value => {\n    if (value !== \"a\") {\n      this.setState({\n        value: value,\n        options: [\"ab\", \"abbbbb\", \"abbbbb\", \"aaaab\"]\n      });\n    } else {\n      this.setState({\n        value: value,\n        options: [\"a\", \"aa\", \"aaa\", \"aaaa\"]\n      });\n    }\n  };\n  render() {\n    let { value, options, placeholder, disabled } = this.state;\n    return (\n      <div className=\"demo\" style={{ marginBottom: \"110px\" }}>\n        <AutoComplete\n          value={value}\n          disabled={disabled}\n          options={options}\n          placeholder={placeholder}\n          onValueChange={value => this.onFormChange(value)}\n        />\n      </div>\n    );\n  }\n}\n\n\n","desc":" 用户可以按需进行选择，支持光标操作，回车事件"}]


class Demo extends Component {
    constructor(props) {
        super(props);
        this.state = {
            open: false
        }
        this.handleClick = this.handleClick.bind(this);
    }

    handleClick() {
        this.setState({open: !this.state.open})
    }

    render() {
        const {title, example, code, desc, scss_code} = this.props;
        let caret = this.state.open ? CARETUP : CARET;
        let text = this.state.open ? "隐藏代码" : "查看代码";

        const header = (
            <div>
                {example}
                <Button style={{"marginTop": "10px"}} shape="block" onClick={this.handleClick}>
                    {caret}
                    {text}
                </Button>
            </div>
        );
        return (
            <Col md={12}>
                <h3>{title}</h3>
                <p>{desc}</p>
                <Panel collapsible headerContent expanded={this.state.open} colors='bordered' header={header}
                       footerStyle={{padding: 0}}>
                    <pre><code className="hljs javascript">{code}</code></pre>
                    {!!scss_code ? <pre><code className="hljs css">{scss_code}</code></pre> : null}
                </Panel>
            </Col>
        )
    }
}

class DemoGroup extends Component {
    constructor(props) {
        super(props)
    }

    render() {
        return (
            <Row>
                {DemoArray.map((child, index) => {

                    return (
                        <Demo example={child.example} title={child.title} code={child.code} scss_code={child.scss_code}
                              desc={child.desc} key={index}/>
                    )

                })}
            </Row>
        )
    }
}

ReactDOM.render(<DemoGroup/>, document.getElementById('tinperBeeDemo'));
